server/src/main/java/org/apache/druid/client/CachingClusteredClient.java

```diff
@@ -812,7 +812,7 @@ String computeResultLevelCachingEtag(
       Hasher hasher = Hashing.sha1().newHasher();
       boolean hasOnlyHistoricalSegments = true;
       for (SegmentServerSelector p : segments) {
-        if (!p.getServer().pick().getServer().isSegmentReplicationTarget()) {
+        if (!p.getServer().pick(null).getServer().isSegmentReplicationTarget()) {
```

Comment:
this was the only functional place that the parameterless pick() was called that I mentioned in the other comment @jihoonson , the Query is available inside of SpecificQueryRunnable where this is called, so I could alternatively pass the query into computeResultLevelCachingEtag and add a null check on this line, what do you think?

Reply:
I made the query change.
For the null check I was thinking for this part p.getServer().pick(query) since we just annotated this Nullable below?
Something like -
QueryableServer queryableServer = p.getServer().pick(query);
if (queryableServer==null || !queryableServer.getServer().isSegmentReplicationTarget()) {

Reply:
That's a good catch! I think we should add a null check.